### PR TITLE
Implementation of the unpooling layer in tf.contrib.layers

### PR DESCRIPTION
**Pull request by rayanelleuch**
This pull request follow the propositions from #2169 
And try to reproduce the layer implemented in https://arxiv.org/abs/1505.04366

I integrated the code as a layer in tf.contrib.layers

There is most probably additional changes that are needed.
I saw that there should be a testing unit added but I don't know exactly how it should be done.

### CODE REVIEW

**Comment by rayanelleuch** [comment]
I signed it!

**Comment by fchollet** [comment]
Yes, we're promoting class-based layers and we will not add such function layers. Also, if it doesn't have a state (weights), then it is better for it not to be a layer at all.

Is "unpooling" as you define it the same as "repeat"? e.g.

```
A B
C D

becomes

A A B B
A A B B
C C D D
```

**Comment by rayanelleuch** [comment]
It is true that it does not have states. So is there an other place where it would fit than in layers?
But I want to point the fact that max_pooling, average_pooling  do not have states and are present in layers.

No it is not as repeat.
unpool serve to resize features map after it have been pooled by tf.nn.max_pool_with_argmax.
The max indices are then used to map the resizing. The other pixels are left at 0. Only the place where the max appeared during the max_pooling will be filled.

Example taken from the paper "Learning Deconvolution Network for Semantic Segmentation"
screenshot from 2018-02-15 12-43-37

An other more numerical example:

```
image = tf.placeholder(tf.float32, [None, 4, 4, 1])
pooled, ind = tf.nn.max_pool_with_argmax(image,[1,2,2,1],[1,2,2,1],"SAME")
unpooled = unpool(pooled, ind)
```

image ->
```
[[4.17022005e-01 7.20324493e-01 1.14374817e-04 3.02332573e-01]
 [1.46755891e-01 9.23385948e-02 1.86260211e-01 3.45560727e-01]
 [3.96767474e-01 5.38816734e-01 4.19194514e-01 6.85219500e-01]
 [2.04452250e-01 8.78117436e-01 2.73875932e-02 6.70467510e-01]]
```

pooled->
```
[[0.7203245  0.34556073]
 [0.87811744 0.6852195 ]]
```

unpooled->
```
[[0.         0.7203245  0.         0.        ]
 [0.         0.         0.         0.34556073]
 [0.         0.         0.         0.6852195 ]
 [0.         0.87811744 0.         0.        ]]
```

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ispirmustafa: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ispirmustafa: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by FengYangTL** [comment]
Any idea which release will this one make into?  I would like to use it in an autoencoder...

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by fchollet** [comment]
Thank you for your explanation, and sorry for the delayed feedback.

Based on the following:

> we're promoting class-based layers and we will not add such function layers
> if it doesn't have a state (weights), then it is better for it not to be a layer at all.

We will not add this layer to tf.contrib.

Thanks for the PR regardless.

**Comment by ebrevdo** [comment]
This may be a good fit for tf.contrib.nn?

**Comment by rayanelleuch** [comment]
@fchollet 
That explanation is not enough. It's contradictory.
I'd like to know why max_pooling or average_pooling are present then.

If it is just about the class-based implementation it can be modified to work.

**Comment by stesha2016** [comment]
About this solution, we need fix batch size. How can we handle with unfixed batch size for unpooling?
